src/NUnitFramework/framework/Constraints/UniqueItemsConstraint.cs

```diff
@@ -52,97 +55,112 @@ public override string Description
+                    processedItems.Add(o1);
+                else if (unknownNonUnique)
+                    nonUniques.Add(o1);
             }
 
-            return true;
+            return nonUniques;
         }
 
-        private bool? TryFastAlgorithm(IEnumerable actual)
+        private ICollection? TryFastAlgorithm(IEnumerable actual)
         {
             // If the user specified any external comparer with Using, exit
             if (UsingExternalComparer)
                 return null;
 
             // If IEnumerable<T> is not implemented exit,
             // Otherwise return value is the Type of T
-            Type memberType = GetGenericTypeArgument(actual);
+            Type? memberType = GetGenericTypeArgument(actual);
             if (memberType == null || !IsSealed(memberType) || IsHandledSpeciallyByNUnit(memberType))
                 return null;
 
             // Special handling for ignore case with strings and chars
             if (IgnoringCase)
             {
                 if (memberType == typeof(string))
-                    return StringsUniqueIgnoringCase((IEnumerable<string>)actual);
+                    return (ICollection)StringsUniqueIgnoringCase((IEnumerable<string>)actual);
                 else if (memberType == typeof(char))
-                    return CharsUniqueIgnoringCase((IEnumerable<char>)actual);
+                    return (ICollection)CharsUniqueIgnoringCase((IEnumerable<char>)actual);
             }
 
-            return (bool)ItemsUniqueMethod.MakeGenericMethod(memberType).Invoke(null, new object[] { actual });
+            return (ICollection)ItemsUniqueMethod.MakeGenericMethod(memberType).Invoke(null, new object[] { actual });
         }
 
-        private bool IsSealed(Type type)
+        private static bool IsSealed(Type type)
         {
             return type.GetTypeInfo().IsSealed;
         }
 
         private static readonly MethodInfo ItemsUniqueMethod =
             typeof(UniqueItemsConstraint).GetMethod(nameof(ItemsUnique), BindingFlags.Static | BindingFlags.NonPublic);
 
-        private static bool ItemsUnique<T>(IEnumerable<T> actual)
-        {
-            var hash = new HashSet<T>();
+        private static readonly CaseInsensitiveCharComparer InsensitiveCharComparer =
+            new CaseInsensitiveCharComparer();
 
-            foreach (T item in actual)
-            {
-                if (!hash.Add(item))
-                    return false;
-            }
-
-            return true;
-        }
+        private static ICollection<T> ItemsUnique<T>(IEnumerable<T> actual)
+            => NonUniqueItemsInternal(actual, EqualityComparer<T>.Default);
 
-        private static bool StringsUniqueIgnoringCase(IEnumerable<string> actual)
-        {
-            var hash = new HashSet<string>();
+        private static ICollection<string> StringsUniqueIgnoringCase(IEnumerable<string> actual)
+            => NonUniqueItemsInternal(actual, StringComparer.CurrentCultureIgnoreCase);
```

Comment:
Nice - our string comparer wasn't very culture friendly before! I think this is a separate bug fixed here, two for the price of one!

Reply:
No problem! To be honest though, I'm not the best at culture stuff. I went with CurrentCulture here since I think that matches the existing behaviour of ToLower(), but I'm unsure if we should instead be using InvariantCulture or Ordinal for the ignore case.

Reply:
CurrentCulture is the right choice here - we need to use CurrentCulture, as users can set the culture on any individual test themselves with the `SetCulture` attribute. 🙂

Reply:
Oh right, of course! :)